src/deploy/functions/prepare.ts

```diff
@@ -78,114 +74,117 @@ export async function prepare(
-  logger.debug(`Building ${runtimeDelegate.name} source`);
-  await runtimeDelegate.build();
+  context.sources = {};
+  const codebaseUsesEnvs = new Set<string>();
+  const wantBackends: Record<string, backend.Backend> = {};
+  for (const codebase of codebases) {
+    logLabeledBullet("functions", `preparing codebase ${clc.bold(codebase)} for deployment`);
 
-  const firebaseEnvs = functionsEnv.loadFirebaseEnvs(firebaseConfig, projectId);
-  const userEnvOpt: functionsEnv.UserEnvsOpts = {
-    functionsSource: sourceDir,
-    projectId: projectId,
-    projectAlias: options.projectAlias,
-  };
-  const userEnvs = functionsEnv.loadUserEnvs(userEnvOpt);
-  const usedDotenv = hasDotenv(userEnvOpt);
-  const tag = hasUserConfig(runtimeConfig)
-    ? usedDotenv
-      ? "mixed"
-      : "runtime_config"
-    : usedDotenv
-    ? "dotenv"
-    : "none";
-  void track("functions_codebase_deploy_env_method", tag);
+    const config = configForCodebase(context.config, codebase);
+    const sourceDirName = config.source;
+    if (!sourceDirName) {
+      throw new FirebaseError(
+        `No functions code detected at default location (./functions), and no functions source defined in firebase.json`
+      );
+    }
+    const sourceDir = options.config.path(sourceDirName);
+    const delegateContext: runtimes.DelegateContext = {
+      projectId,
+      sourceDir,
+      projectDir: options.config.projectDir,
+      runtime: config.runtime || "",
+    };
+    const runtimeDelegate = await runtimes.getRuntimeDelegate(delegateContext);
+    logger.debug(`Validating ${runtimeDelegate.name} source`);
+    await runtimeDelegate.validate();
+    logger.debug(`Building ${runtimeDelegate.name} source`);
+    await runtimeDelegate.build();
 
-  logger.debug(`Analyzing ${runtimeDelegate.name} backend spec`);
-  const wantBackend = await runtimeDelegate.discoverSpec(runtimeConfig, firebaseEnvs);
-  wantBackend.environmentVariables = { ...userEnvs, ...firebaseEnvs };
-  for (const endpoint of backend.allEndpoints(wantBackend)) {
-    endpoint.environmentVariables = wantBackend.environmentVariables;
-    endpoint.codebase = context.config.codebase;
+    const firebaseEnvs = functionsEnv.loadFirebaseEnvs(firebaseConfig, projectId);
+    const userEnvOpt: functionsEnv.UserEnvsOpts = {
+      functionsSource: sourceDir,
+      projectId: projectId,
+      projectAlias: options.projectAlias,
+    };
+    const userEnvs = functionsEnv.loadUserEnvs(userEnvOpt);
+    logger.debug(`Analyzing ${runtimeDelegate.name} backend spec`);
+    const wantBackend = await runtimeDelegate.discoverSpec(runtimeConfig, firebaseEnvs);
+    wantBackend.environmentVariables = { ...userEnvs, ...firebaseEnvs };
+    for (const endpoint of backend.allEndpoints(wantBackend)) {
+      endpoint.environmentVariables = wantBackend.environmentVariables;
+      endpoint.codebase = codebase;
+    }
+    wantBackends[codebase] = wantBackend;
+    if (functionsEnv.hasUserEnvs(userEnvOpt)) {
```

Comment:
we can add `&& !codebaseUsesEnvs.includes(codebase)` to keep duplicates out if we switch to using an array

Reply:
At this point, we are guaranteed to have unique codebases, so I'll keep that out for brevity!